Porting the Pooling kernel to Tensorflow Lite Micro with MaxPool support

jenselofsson opened the pull request:
This PR ports the MaxPool-part of the pooling kernel from Tensorflow Lite to Tensorflow Lite Micro.

jenselofsson committed:
Add pooling kernel to Tensorflow Lite Micro.

jenselofsson committed:
Add testcases for floating point MaxPooling.

The testcases are copied from the Tensorflow Lite testcases for MaxPool.

jenselofsson committed:
Add test-function for MaxPoolFloat.

jenselofsson committed:
Add test function for uint8 MaxPooling.

jenselofsson committed:
Add testcases for uint8 MaxPooling.

The testcases are copied from Tensorflow Lite.

jenselofsson committed:
Add pooling to the BUILD file.

wangtz commented on tensorflow/lite/experimental/micro/kernels/all_ops_resolver.cc:
```diff
@@ -29,12 +29,18 @@ TfLiteRegistration* Micro_Register_FULLY_CONNECTED() {
 TfLiteRegistration* Micro_Register_SOFTMAX() { return Register_SOFTMAX(); }
 
+TfLiteRegistration* Register_MAX_POOL_REF();
+TfLiteRegistration* Micro_Register_MAX_POOL_REF() {
```
Since other opts don't have _REF suffix for REF impl. Probably worth keeping it consistent.

wangtz commented on tensorflow/lite/experimental/micro/kernels/pooling.cc:
```diff
@@ -0,0 +1,152 @@
+#include "tensorflow/lite/kernels/internal/reference/pooling.h"
+#include "tensorflow/lite/c/builtin_op_data.h"
+#include "tensorflow/lite/c/c_api_internal.h"
+#include "tensorflow/lite/kernels/internal/reference/integer_ops/pooling.h"
```
Not very clear why you need integer_ops/pooling.h These files are for int8 quantization, and I think we're looking into uint8 version which is more stable.

jenselo replied:
It seems that integer_ops/pooling.h got left behind when I was removing some other code. I'll remove that include.

wangtz commented on tensorflow/lite/experimental/micro/kernels/BUILD:
```diff
@@ -50,6 +51,7 @@ cc_library(
     name = "portable_optimized_micro_ops",
     srcs = [
         "fully_connected.cc",
+        "pooling.cc",
```
Do we really have an optimized version already?

jenselo replied:
No, that line made it in by mistake, I'll remove it.

wangtz reviewed (commented):
Looks great. Thanks for your work. Please see comments below.

jenselofsson committed:
Fix review comments.

Remove pooling.cc from the BUILD file that was added incorrectly.
Remove the _REF from Register_MAX_POOL_REF.
Don't include .../internal/reference/integer_ops/pooling.h file.

wangtz commented:
Hi Jens,

I've added pooling.cc for avg_pool. Please merge. 

Thanks.

jenselo commented:
The force-push is due to rebasing PR onto master, and no code have been added.

wangtz commented on tensorflow/lite/experimental/micro/kernels/all_ops_resolver.cc:
```diff
@@ -34,12 +34,18 @@ TfLiteRegistration* Micro_Register_AVERAGE_POOL_2D() {
 
+TfLiteRegistration* Register_MAX_POOL();
+TfLiteRegistration* Micro_Register_MAX_POOL() {
+  return Register_MAX_POOL();
```
MAX_POOL_2D

jenselo replied:
Done.

wangtz reviewed (commented):
A few nits.

jenselofsson committed:
Fix review comments.

Rename MAX_POOL to MAX_POOL_2D in order to be consistent.
Use TfLiteTypeGetName for printing input type in error message.

jenselo commented on tensorflow/lite/experimental/micro/kernels/all_ops_resolver.cc:
```diff
@@ -34,12 +34,18 @@ TfLiteRegistration* Micro_Register_AVERAGE_POOL_2D() {
   AddBuiltin(BuiltinOperator_FULLY_CONNECTED, Micro_Register_FULLY_CONNECTED(),
              /* min_version */ 1,
              /* max_version */ 2);
+  AddBuiltin(BuiltinOperator_MAX_POOL_2D, Micro_Register_MAX_POOL());
```
Done.

jenselo commented on tensorflow/lite/experimental/micro/kernels/pooling.cc:
```diff
@@ -130,6 +172,30 @@ TfLiteStatus AverageEval(TfLiteContext* context, TfLiteNode* node) {
+      break;
+    default:
+      context->ReportError(context, "Type %d not currently supported.",
+                           input->type);
```
Done.

wangtz reviewed (dismissed):
LGTM

gbaned commented:
@jenselo can you please check build failures. Thanks !

jenselo commented:
> @jenselo can you please check build failures. Thanks !

Yes, I'll take a look.

jenselo commented:
It seems that the test case that fails is the testStableName in the //tensorflow/python:function_test tests (or function_test_gpu in the Linux GPU build).

For the Linux GPU build the batch_matmul_test in //tensorflow/python/compiler/tensorrt.

I've also ran the function_test locally with the maxpool patches and didn't see the error.

Do changes made in tensorflow/lite/experimental/micro have any impact to the tests in tensorflow/python?

jenselofsson committed:
Add pooling.cc to the portable_optimized_micro_ops in the BUILD file.

jenselofsson committed:
Correct BUILD-file with buildifier.
